sympy/core/basic.py

```diff
@@ -1485,7 +1487,7 @@ def rec_replace(expr):
             result = _query(expr)
             if result or result == {}:
                 v = _value(expr, result)
-                if v is not None and v != expr:
+                if v is not None and not _aresame(v, expr):
```

Comment:
Is the only difference between == and _aresame the way they compare floats? I would have expected the two to have roughly the same algorithm, with == working recursively and _aresame using preorder_traversal, and hence roughly the same performance. We should profile _aresame to see why it is so slow.

Although I also agree with Oscar that we should just make == work like this so that _aresame isn't needed. This sort of change implies that this sort of thing will be "wrong" for any code that uses == instead of _aresame, which implies that that's really the behavior that should be in ==.

Reply:
I guess == uses the hashable_content, which is basically directly comparing nested tuples of args instead of doing a preorder traversal. So it's the difference between a C loop and a pure Python loop. I'm somewhat doubtful we can make _aresame as fast as ==, although clearly any micro-optimization we can make on it will have an impact on performance as long as it is used in places like this. 

IMHO we should revert this and mark #19976 as blocked on #20033.